### ISSUE TITLE
No resource identifier found for attribute 'autoVerify' in package 'android'

### ISSUE THREAD

**Issue by tkuben** [BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK#50]
When I build the testbed app, I am getting the following error. 

```

FAILED



FAILURE: Build failed with an exception.


* What went wrong:

Execution failed for task ':processDebugResources'.

> 
com.android.ide.common.internal.LoggedErrorException: Failed to run command:

  	/Users/tkuben/Library/Android/sdk/build-tools/23.0.2/aapt package -f --no-crunch -I /Users/tkuben/Library/Android/sdk/platforms/android-22/android.jar -M /Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml -S /Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/intermediates/res/debug -A /Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/intermediates/assets/debug -m -J /Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/generated/source/r/debug -F /Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/intermediates/res/resources-debug.ap_ --debug-mode --custom-package io.branch.testbed -0 apk --output-text-symbols /Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/intermediates/symbols/debug

  Error Code:

  	1

  Output:

  	/Users/tkuben/Documents/webroot/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/testbed/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml:41: error: No resource identifier found for attribute 'autoVerify' in package 'android'

  	





* Try:

Run with --stacktrace option to get the stack trace. Run with --info or --debug
 option to get more log output.


BUILD FAILED



Total time: 11.1 secs

```

Any Thoughts? 
Cheers, 
TJ

**Comment by aaustin** [comment]
@tkuben Thanks for the heads up - I'm poking a couple engineers who can take a look. Thanks!

**Comment by tkuben** [comment]
I've updated the build.gradle from: 

`
defaultConfig {
        versionCode cdvVersionCode ?: Integer.parseInt("" + privateHelpers.extractIntFromManifest("versionCode") + "0")
        if (cdvMinSdkVersion != null) {
            minSdkVersion cdvMinSdkVersion
        }
    }

    compileSdkVersion cdvCompileSdkVersion
    buildToolsVersion cdvBuildToolsVersion
`

To 

`
    defaultConfig {
        if (cdvMinSdkVersion != null) {
            minSdkVersion cdvMinSdkVersion
        }
        minSdkVersion 19
        targetSdkVersion 23
        versionCode 1
        versionName "1.0"
    }
    compileSdkVersion 23
    buildToolsVersion '23.0.1'
`

With this I am getting the following errors: 

`
:compileDebugJava
/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:776: error: cannot find symbol

                PluginResult result = new PluginResult(PluginResult.Status.OK);

                ^

  symbol:   class PluginResult

  location: class BranchSDK.ShowShareSheetListener

/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:776: error: cannot find symbol

                PluginResult result = new PluginResult(PluginResult.Status.OK);

                                          ^

  symbol:   class PluginResult

  location: class BranchSDK.ShowShareSheetListener

/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:776: error: package PluginResult does not exist

                PluginResult result = new PluginResult(PluginResult.Status.OK);

                                                                   ^

/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:790: error: cannot find symbol

                PluginResult result = new PluginResult(PluginResult.Status.OK);

                ^
... 
`

Ideally I do not want to set my target SDK Version to 6.0. I would like to keep it at 5.0. Your help would be much appreciated. 

Cheers, 
TJ

**Comment by MathildeLemee** [comment]
I've got the same error :( So can't use the new plugin on android.

```
 symbol:   class PluginResult
  location: class BranchSDK.ShowShareSheetListener
/Users/L/code/magic/platforms/android/src/io/branch/BranchSDK.java:861: error: package PluginResult does not exist
                PluginResult result = new PluginResult(PluginResult.Status.OK, response);

```

**Comment by Elpetro** [comment]
Same problem...
Trying to revert to 1.8.0 fails also :+1: 

> $ cordova plugin add io.branch.sdk@1.8.0
> Fetching plugin "io.branch.sdk@1.8.0" via npm
> Fetching plugin "io.branch.sdk@1.8.0" via cordova plugins registry
> Error: Failed to fetch plugin io.branch.sdk@1.8.0 via registry.
> Probably this is either a connection problem, or plugin spec is incorrect.
> Check your connection and plugin name/version/URL.
> Error: 404 Not Found: io.branch.sdk

**Comment by olliebennett** [comment]
Also have this problem. Seeing the exact same error around `PluginResult`, and likewise cannot downgrade to an older version.

Have tried with cordova versions `6.0.0` and `5.4.1`.

```
$ cordova plugin list
cordova-plugin-android-support-v4 21.0.1 "Android Support v4"
cordova-plugin-app-version 0.1.8 "AppVersion"
cordova-plugin-camera 2.1.0 "Camera"
cordova-plugin-console 1.0.2 "Console"
cordova-plugin-device 1.1.1 "Device"
cordova-plugin-facebook4 1.0.0 "Facebook Connect"
cordova-plugin-file 4.1.1 "File"
cordova-plugin-file-transfer 1.5.0 "File Transfer"
cordova-plugin-globalization 1.0.2 "Globalization"
cordova-plugin-inappbrowser 1.2.1 "InAppBrowser"
cordova-plugin-splashscreen 3.1.0 "Splashscreen"
cordova-plugin-statusbar 2.1.1 "StatusBar"
cordova-plugin-whitelist 1.2.1 "Whitelist"
cordova-plugin-x-socialsharing 5.0.10 "SocialSharing"
io.branch.sdk 1.0.0 "branch-cordova-sdk"
ionic-plugin-keyboard 1.0.8 "Keyboard"
pushbots-cordova-plugin 1.3.5 "Pushbots"
```

```
$ cordova platform list
Installed platforms: android 5.1.1, ios 4.0.1
```

Installing `cordova-plugin-crosswalk-webview` also makes no difference.

Currently having to remove branch altogether in order to get a functioning build together!

**Comment by olliebennett** [comment]
Update: I've "downgraded" for now by explicitly defining the git commit hash (b89bb4a9bf2e0eca7c9f9a51dae7c61a61c15e0f) for the commit prior to this branch being merged:

```
cordova plugin add https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK#b89bb4a9bf2e0eca7c9f9a51dae7c61a61c15e0f --variable BRANCH_LIVE_KEY="ABC" --variable URI_SCHEME="splittable" 
```

**Comment by aaustin** [comment]
Hey team - sorry about this. I've alerted the engineers to take a look but they didn't get a chance. Does it seem like the root cause is the auto verify stuff? I'll ask them to remove it.

**Comment by tkuben** [comment]
Hi @aaustin 

Unfortunately, the auto verify does not seem to be the issue regarding the errors with PluginResult. It is an issue that we have to use the latest SDK to compile, however we still have compile issues with the autoverify removed. I modified the android.json & AndroidManifest.xml to remove the autoVerify="true"  and I am still getting... 

```
:compileDebugJava
/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:776: error: cannot find symbol

                PluginResult result = new PluginResult(PluginResult.Status.OK);

                ^
  symbol:   class PluginResult
  location: class BranchSDK.ShowShareSheetListener

/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:776: error: cannot find symbol

                PluginResult result = new PluginResult(PluginResult.Status.OK);

                                          ^

  symbol:   class PluginResult

  location: class BranchSDK.ShowShareSheetListener

/Users/tkuben/Documents/webroot/riipe/mobile/platforms/android/src/io/branch/BranchSDK.java:776: error: package PluginResult does not exist

                PluginResult result = new PluginResult(PluginResult.Status.OK);
....
```

Your help in resolving this issue would be much appreciated. 

Cheers, 
TJ

**Comment by tkuben** [comment]
Hi @aaustin 

The issue is caused by the fact that in: ./src/android/io/branch/BranchSDK.java, PluginResult from cordovaLib is referenced, however is not imported. Please commit the following changeset to your codebase. I tried to create a new branch on your repo, however I got permission denied. If you grant me access I can create a branch and create a pull request for you. 

```
diff --git a/src/android/io/branch/BranchSDK.java b/src/android/io/branch/BranchSDK.java
index a26bf31..5b6612d 100644
--- a/src/android/io/branch/BranchSDK.java
+++ b/src/android/io/branch/BranchSDK.java
@@ -6,6 +6,7 @@ import android.util.Log;
 
 import org.apache.cordova.CallbackContext;
 import org.apache.cordova.CordovaPlugin;
+import org.apache.cordova.PluginResult;
 
 import org.json.JSONArray;
 import org.json.JSONException;
```

We also do need to remove the autoVerify as this is causing issues with setting our targetSDK to 6.0 which is not good as a requirement. 

Cheers, 
TJ

**Comment by ZackMattor** [comment]
I'm testing @tkuben's solutions and will create a PR if it works.

**Comment by ZackMattor** [comment]
@tkuben, i'm still getting `processDebugResources FAILED` ... did you get past that with your gradle changes?

**Comment by tkuben** [comment]
@ZackMattor What is the Failed message you are getting? Yes I did get past with my original build.gradle that was generated using phonegap: 
```
defaultConfig {
        versionCode cdvVersionCode ?: Integer.parseInt("" + privateHelpers.extractIntFromManifest("versionCode") + "0")
        if (cdvMinSdkVersion != null) {
            minSdkVersion cdvMinSdkVersion
        }
    }

    compileSdkVersion cdvCompileSdkVersion
    buildToolsVersion cdvBuildToolsVersion
```

However, I did need to modify my AndroidManifest.xml & android.json to remove the android:autoVerify="true"

Can you post the entire Failed message please? 

Cheers, 
TJ

**Comment by ZackMattor** [comment]
@tkuben Yup.. it's requiring an android SDK version 23+ ... this is quite a problem considering that's solely for android 6.0+ ;)

**Comment by tkuben** [comment]
Exactly @ZackMattor @aaustin. We cannot tie our app just to 6.0+ so thats why I removed the autoverify. Were you able to build, run & test after removing the autoverify and applying my patch? 

Cheers, 
TJ

**Comment by ZackMattor** [comment]
I didn't have to remove autoverify to build.. just bumped up the built sdk... it still runs on the android 5.x devices... I don't have a 4.x phone to test it on...

Let me make sure we still have branch functionality in 5.x, just because it builds and runs.. doesn't always mean it works.

**Comment by ZackMattor** [comment]
@tkuben is there anyway to set the SDK version in config.xml that you know of?.... The only reference for that i see is for phonegap and doesn't seem to be working...

**Comment by gabaum10** [comment]
You can set it like this in the config.xml file:

```
<preference name="android-minSdkVersion" value="19" /> 
<preference name="android-targetSdkVersion" value="21" />
```

**Comment by jestoniyap** [comment]
Hello @tkuben and @ZackMattor, I'm here to help you guys with the issue. Can you give me the summary of the details of the issue you are experiencing as well as the environment you have right now in your machine (npm version, cordova version, etc) so we can test it here on our end? Thank you for your patience.

**Comment by gabaum10** [comment]
So I'm having the same issue too.  Here is my setup.

-building through phonegap build
-cli-6.0.0
-npm plugin version 2.0
-android target api:21
-android min api:14

I haven't had the time to set up a local environment yet, but the pgb servers areally getting the same error outlined here with this setup.

**Comment by gabaum10** [comment]
Also, interestingly, as mentioned in #51 if you build out of npm, you won't get this error, but iOS will fail.  If you build out of master, iOS will build but android will fail.

**Comment by ZackMattor** [comment]
On master...
```
npm -v
2.12.1

cordova -v
5.4.1
```

```
    <preference name="android-minSdkVersion" value="19" />
    <preference name="android-targetSdkVersion" value="23"/>
```

`cordova build android --stacktrace --debug`
```:mergeDebugAssets UP-TO-DATE
:generateDebugResValues UP-TO-DATE
:generateDebugResources UP-TO-DATE
:mergeDebugResources UP-TO-DATE
:processDebugManifest UP-TO-DATE
:processDebugResources
/Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml:72: error: No resource identifier found for attribute 'autoVerify' in package 'android'

:processDebugResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processDebugResources'.
> com.android.ide.common.internal.LoggedErrorException: Failed to run command:
        /Users/zmattor/Library/Android/sdk/build-tools/23.0.2/aapt package -f --no-crunch -I /Users/zmattor/Library/Android/sdk/platforms/android-22/android.jar -M /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml -S /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/res/debug -A /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/assets/debug -m -J /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/generated/source/r/debug -F /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/res/resources-debug.ap_ --debug-mode --custom-package com.getadored.adored -0 apk --output-text-symbols /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/symbols/debug
  Error Code:
        1
  Output:
        /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build/intermediates/manifests/full/debug/AndroidManifest.xml:72: error: No resource identifier found for attribute 'autoVerify' in package 'android'



* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 2.958 secs

/Users/zmattor/work/adored/adored-ember/cordova/platforms/android/cordova/node_modules/q/q.js:126
                    throw e;
                          ^
Error code 1 for command: /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/gradlew with args: cdvBuildDebug,-b,/Users/zmattor/work/adored/adored-ember/cordova/platforms/android/build.gradle,-Dorg.gradle.daemon=true
ERROR building one of the platforms: Error: /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/cordova/build: Command failed with exit code 1
You may not have the required environment or OS to build this project
Error: /Users/zmattor/work/adored/adored-ember/cordova/platforms/android/cordova/build: Command failed with exit code 1```

**Comment by renesansz** [comment]
@gabaum10 
I've tried installing the plugin thru `phonegap` and I only get the `PluginResult` error, however I'm going to submit a fix later on for this..
However I can't replicate other issues aside from what you've specified.

Could you try building it out locally first?.
Make sure to just modify partially the missing import `PluginResult` in Java source code.

Also, could you comment your current `config.xml`? Thanks

**Comment by gabaum10** [comment]
@renesanz, you can specify to build out of the git master if you set the "value" property to the git url.  If you do that the PluginResult issue is indeed fixed, and you'll see the issue I'm talking about.

I'll post my config when I get to a computer.

**Comment by gabaum10** [comment]
@renesansz it looks like the only error I'm still getting in master is the `PluginResult` error.  Once that is merged in, it looks like everything should work.  iOS is building correctly.

**Comment by aaustin** [comment]
@gabaum10 I pushed the fix for this to NPM. Please confirm it's resolved?

**Comment by ZackMattor** [comment]
@aaustin, we were playing with the PR before it got merged, and it seems to be working! Thank you! :+1:

We will test the NPM package soon, but i'm sure it will work.

**Comment by aaustin** [comment]
Thanks for your feedback guys :)

**Comment by gabaum10** [comment]
This is working for me.

**Comment by madc0w** [comment]
It looks like the latest version of the Cordova plugin is still 2.0.2:

https://cordova.apache.org/plugins/?q=branch

Can you give some indication of when this fix will be pushed to the Cordova plugin repo?

**Comment by aaustin** [comment]
@madc0w deployed 2.0.3 to NPM - please give it a try!

**Comment by madc0w** [comment]
I am using Cordova via Meteor.  Adding an NPM module to Meteor seems quite complicated... Adding a plugin from the standard Cordova repo is easy, though.

**Comment by madc0w** [comment]
I've tried using the meteorhacks:npm package to add this NPM module.  Unfortunately, this package only allows access to the NPM module on the server side, whereas it seems to make use of "window", requiring client-side execution.  So... I tried using browserify (see https://react-in-meteor.readthedocs.org/en/latest/client-npm/ )... but that was no help, either:

W20160314-16:14:34.367(1)? (STDERR) ReferenceError: window is not defined
W20160314-16:14:34.368(1)? (STDERR)     at node_modules/branch-sdk/dist/build.min.js:5:1
W20160314-16:14:34.368(1)? (STDERR)     at Object.<anonymous> (node_modules/branch-sdk/dist/build.min.js:78:1)
W20160314-16:14:34.368(1)? (STDERR)     at s (../../../../../.meteor/packages/cosmos_browserify/.0.10.0.zzcf9t++os+web.browser+web.cordova/plugin.CosmosBrowserify.os/npm/CosmosBrowserify/node_modules/browserify/node_modules/browser-pack/_prelude.js:1:1)
W20160314-16:14:34.368(1)? (STDERR)     at ../../../../../.meteor/packages/cosmos_browserify/.0.10.0.zzcf9t++os+web.browser+web.cordova/plugin.CosmosBrowserify.os/npm/CosmosBrowserify/node_modules/browserify/node_modules/browser-pack/_prelude.js:1:1
W20160314-16:14:34.368(1)? (STDERR)     at Object.branch-sdk (_stream_0.js:1:1)
W20160314-16:14:34.369(1)? (STDERR)     at s (../../../../../.meteor/packages/cosmos_browserify/.0.10.0.zzcf9t++os+web.browser+web.cordova/plugin.CosmosBrowserify.os/npm/CosmosBrowserify/node_modules/browserify/node_modules/browser-pack/_prelude.js:1:1)
W20160314-16:14:34.369(1)? (STDERR)     at e (../../../../../.meteor/packages/cosmos_browserify/.0.10.0.zzcf9t++os+web.browser+web.cordova/plugin.CosmosBrowserify.os/npm/CosmosBrowserify/node_modules/browserify/node_modules/browser-pack/_prelude.js:1:1)
W20160314-16:14:34.369(1)? (STDERR)     at ../../../../../.meteor/packages/cosmos_browserify/.0.10.0.zzcf9t++os+web.browser+web.cordova/plugin.CosmosBrowserify.os/npm/CosmosBrowserify/node_modules/browserify/node_modules/browser-pack/_prelude.js:1:1
W20160314-16:14:34.369(1)? (STDERR)     at node_modules/branch-sdk/dist/build.min.js:79:1
W20160314-16:14:34.369(1)? (STDERR)     at /home/mad/spoticle/.meteor/local/build/programs/server/boot.js:242:10


I see that the version in the NPM repo 2.1.1.  The version in the Cordova repo is currently 2.0.2.  Can we please have an update in the Cordova repo?

**Comment by madc0w** [comment]
It seems I spoke too soon.  A Cordova plugin version 2.0.3 of branch-cordova-sdk is available from the Crodova repo.  It requires many NPM modules, which must be added manually.  After adding meteorhacks:npm, I added the following to packages.json:

{
	"mkpath": "1.0.0",
	"node-version-compare": "1.0.1",
	"plist": "1.2.0",
	"rimraf": "2.5.2",
	"xml2js": "0.4.16"
}

And now, I get:

npm-container: updating npm dependencies -- mkdirp, mkpath, node-version-compare, plist, rimraf, xml2js...
Checking dependencies:plugins                [=====================      ] 76% 7.8s
    Node module mkpath is found
    Module mkpath is installed
    Node module node-version-compare is found
    Module node-version-compare is installed
    Node module plist is found
    Module plist is installed
    Node module rimraf is found
    Module rimraf is installed
    Node module xml2js is found
    Module xml2js is installed
% Error during processing of action! Attempting to revert...                       
% Failed to install 'io.branch.sdk':Error: Uh oh!                                  
**"/home/mad/spoticle/.meteor/local/cordova-build/plugins/io.branch.sdk/src/ios/dependencies/Branch-SDK/BNCContentDiscoveryManager.m" not found!
**  at Object.module.exports.common.copyFile (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/platforms/common.js:38:40)
  at Object.module.exports.common.copyNewFile (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/platforms/common.js:69:16)
  at installHelper (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/platforms/ios.js:54:16)
  at Object.module.exports.source-file.install (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/platforms/ios.js:106:13)
  at installWrapper (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/platforms/platforms.js:77:32)
  at Object.ActionStack.process (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/util/action-stack.js:68:25)
  at handleInstall (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/install.js:576:20)
  at /home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/cordova-lib/src/plugman/install.js:363:28
  at _fulfilled (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/q/q.js:787:54)
  at self.promiseDispatch.done (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/q/q.js:816:30)
  at Promise.promise.promiseDispatch (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/q/q.js:749:13)
  at /home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/q/q.js:509:49
  at flush (/home/mad/.meteor/packages/meteor-tool/.1.1.10.1aaar8u++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/q/q.js:108:17)
  at process._tickCallback (node.js:448:13)

=> Errors executing Cordova commands:                                              
                                                                                   
   While adding plugin branch-cordova-sdk@2.0.3 to Cordova project:                
   Error: Uh oh!

**Comment by gabaum10** [comment]
Known issue, see https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/issues/71

**Comment by madc0w** [comment]
Now with 2.0.5, I am able to get a little bit farther.

As per the docs here https://dev.branch.io/getting-started/sdk-integration-guide/guide/cordova/#start-a-branch-session , I attempt to init the session.  Branch.initSession() returns without error, but then I see:

I20160315-17:04:10.717(1) (android:http://meteor.local/plugins/io.branch.sdk/www/branch.js:27) Uncaught (in promise) Attempt to invoke virtual method 'android.content.Context android.app.Activity.getApplicationContext()' on a null object reference

Later, when I attempt to generate a link using code from the docs here https://dev.branch.io/getting-started/creating-links-in-apps/guide/cordova/ : 

			var branchUniversalObj = null;
			Branch.createBranchUniversalObject({
				canonicalIdentifier : comment._id,
				title : comment.headline,
				contentDescription : comment.text,
				contentImageUrl : image,
				contentMetadata : {
					userId : Meteor.user() ? Meteor.user()._id : null,
					userName : Meteor.user() ? Meteor.user().username : null
				}
			}).then(function(newBranchUniversalObj) {
				branchUniversalObj = newBranchUniversalObj;
			});

I see:

I20160315-17:08:46.560(1) (android:http://meteor.local/:0) Uncaught (in promise) Invalid action

The "then" function is never called, so branchUniversalObj == null.

**Comment by madc0w** [comment]
I'm back with a bit more information.

Shortly after startup, I see:

<branch-config> tag is not set in the config.xml. Universal Links plugin is not going to work.
   
So... that can't be good.

Also, attempting to initSession in the "resume" handler results in the behavior described above.  If I do the initSession in the "deviceready" handler only, then I (sometimes!) see a different error message:

Uncaught (in promise) ReferenceError: DeepLinkHandler is not defined

I have yet to isolate the exact conditions which determine which error will be produce.  Nonetheless, when this happens, then I am finally able to generate the link!  The code given in the  docs here https://dev.branch.io/getting-started/creating-links-in-apps/guide/cordova/ is, however, wrong.  res.generatedUrl is undefined.  The generated URL seems to be instead in res.url.

**Comment by asafco4** [comment]
@madc0w , I get the same error, please update if you have a solution...
It appears even though that function exists!

// Init Branch
document.addEventListener('resume', onResume, false);
document.addEventListener('deviceready', onDeviceReady, false);
		
function onResume() {
  console.log("Branch onResume()");
  Branch.initSession();
  console.log("Branch Resumed...");
}
		
function onDeviceReady() {
  console.log("Branch onDeviceReady()");
  Branch.initSession();
  console.log("Branch onDeviceReady ended...");
}
		
function DeepLinkHandler(data) {
  console.log("Branch DeepLinkHandler: Activated");
}

**Comment by madc0w** [comment]
Having received no reply to this thread, we finally gave up on Branch (after a solid week of hair-tearing) and just wrote the deep linking functionality ourselves.  I can only suggest you do the same.

**Comment by asafco4** [comment]
@madc0w , thanks for your reply.
Are you working on Ionic/Cordova or native app?
Isn't writing it yourself very complicated? I know that Android has a built-in way to track referred installations but how did you do it on IOS?

Thanks!

**Comment by aaustin** [comment]
Hey guys - I'm really sorry about the lack of response. We should have re-opened this issue for it to get attention. I'm re-opening it now and forwarding it to the engineers.

**Comment by asafco4** [comment]
@aaustin , thanks for your attention.
Please make sure that the re-open refer to the 2 issues both madc0w and I have:
1. Warning message: < branch-config > tag is not set in the config.xml. Universal Links plugin is not going to work.
2. Error message about missing function even though it's there!
Uncaught (in promise) ReferenceError: DeepLinkHandler is not defined
Cheers!

**Comment by jestoniyap** [comment]
Hello @asafco4, we'll be taking a look at the second issue.
The first is just a warning message if you didn't set the _branch-config_ tag in your config.xml. It is up to you if you want to support Universal Links for your app.

**Comment by renesansz** [comment]
@madc0w @asafco4 ,

We haven't tested the plugin yet in Meteor JS, however in regards to the `<branch-config>` it should automatically append the config in the manifest file. I think we need a separate hook for Meteor?

Also, can you provide us the details of your setup? (eg: version #, platform, etc.).
And also a step-by-step procedure in producing the bug.

Cheers,
Rene

**Comment by asafco4** [comment]
@renesansz , thanks for the quick reply.

I'm kinda new, don't know what you mean by Meteor JS.

I have an app on Ionic/Cordova platform, the instructions I used are from:
https://dev.branch.io/getting-started/sdk-integration-guide/guide/cordova/#start-a-branch-session

The first error is there every time the app is running.

My code is exactly like in the instructions example:
// Init Branch
document.addEventListener('resume', onResume, false);
document.addEventListener('deviceready', onDeviceReady, false);

function onResume() {
console.log("Branch onResume()");
Branch.initSession();
console.log("Branch Resumed...");
}

function onDeviceReady() {
console.log("Branch onDeviceReady()");
Branch.initSession();
console.log("Branch onDeviceReady ended...");
}

function DeepLinkHandler(data) {
console.log("Branch DeepLinkHandler: Activated");
}

I run the app, my app.js file code is loading, branch init session is called when the login page is ready, and I can see that error in the console.
The annoying thing is that function exists (as you can see).
Also, that worked for me couple days ago, didn't change code, just added IOS support for another plugin, not sure why would it affect it.
Cheers!

**Comment by renesansz** [comment]
@asafco4, no worries! :+1: 

However, I'm trying to replicate the issue and I was not able to reproduce the error: `Uncaught (in promise) ReferenceError: DeepLinkHandler is not defined`

![](http://i.imgur.com/bjd4bcl.png)

I'm currently working on with the unhandled promise error (though it won't affect any functionalities of the app, I'm just making sure that all errors are handled correctly)

Would like to confirm if you're using the latest version of Branch Cordova?

**Comment by asafco4** [comment]
Sure,  please instruct me how to verify it.

**Comment by renesansz** [comment]
cool, first let's verify if we're using the same versions of the plugin.

Can you post here the details?

Do the following:
```
cordova plugin -l
npm -v
cordova -v
```

And post here the details.

**Comment by asafco4** [comment]
cordova plugin -l:

cordova-plugin-console 1.0.2 "Console"
cordova-plugin-customurlscheme 4.1.5 "Custom URL scheme"
cordova-plugin-device 1.1.1 "Device"
cordova-plugin-email-composer 0.8.3 "EmailComposer"
cordova-plugin-file 4.1.1 "File"
cordova-plugin-geolocation 2.1.0 "Geolocation"
cordova-plugin-google-analytics 0.8.1 "Google Universal Analytics Plugin"
cordova-plugin-inappbrowser 1.3.0 "InAppBrowser"
cordova-plugin-media 2.2.0 "Media"
cordova-plugin-network-information 1.2.0 "Network Information"
cordova-plugin-splashscreen 3.2.1 "Splashscreen"
cordova-plugin-statusbar 2.1.2 "StatusBar"
cordova-plugin-whitelist 1.2.1 "Whitelist"
cordova-plugin-x-socialsharing 5.0.12-dev "SocialSharing"
es6-promise-plugin 3.0.2 "Promise"
io.branch.sdk 2.0.8 "branch-cordova-sdk"
ionic-plugin-keyboard 2.0.1 "Keyboard"
org.loicknuchel.cordova.deviceaccounts 0.0.1 "Device Accounts"
phonegap-plugin-push 1.6.1 "PushPlugin"


┌──────────────────────────────────────────┐
│ Update available: 6.1.1 (current: 5.4.0) │
│ Run npm install -g cordova to update.    │
└──────────────────────────────────────────┘


npm -v:

3.3.6


cordova -v:

5.4.0

I see I need to update cordova...

**Comment by renesansz** [comment]
Please do.

If the problem still persist could you do the following:
1. Remove the platforms, `cordova platform rm <ios/android>`
2. Remove the plugin `cordova plugin rm io.branch.sdk`
3. Re-install the platform, `cordova platform add <ios/android>`
4. Re-install the plugin, refer to the [instructions](https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK#command-line-install)

Let me know if still doesn't work.

**Comment by asafco4** [comment]
still not working :(
Maybe we can get on Skype or something to share my code with you?

**Comment by asafco4** [comment]
I get the following debug message:
Branch onDeviceReady()
Branch onDeviceReady ended...

but still has that damn error later:
"Uncaught (in promise) ReferenceError: DeepLinkHandler is not defined"

Maybe it expects to have that function in a different scope? maybe the
function should be global or something?
Why is it even called? there's no branch data sent or used currently.

Please assist, I'm getting closer to just give up :(

**Comment by renesansz** [comment]
@asafco4 ,
I don't know if I'm allowed to do that, since I'm just a contractor for this project.
At the moment I can only assist you through replying to this issue.
An alternative would be creating a video in producing the bug (make sure to show and explain the project structure, the command line path, the codes, basically everything).

The reason why `DeepLinkHandler` is being called is due to `initSession`. See this [code](https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/blob/master/src/android/io/branch/BranchSDK.java#L802).

I know this could be silly, but sometimes I do commit this kind of mistake...
First is try to modify your Javascript file and try to put any code that would indicate that it was being called, let's say just an `alert` method (make sure it is being executed before the `uncaught error`). See if the changes reflects upon re-building the app.

**Comment by tkuben** [comment]
Hi @asafco4 

It probably is a scope issue. You should inspect all the JS files the app is trying to load. Particularly look to see if a file called branch.js is being loaded from the root of your application. If so, can you try touching (a blank file) a file at the root of your application called branch.js and try loading the app again? 

Hope it helps. 
Cheers, 
TJ

**Comment by asafco4** [comment]
Sorry, I don't understand what you want me to try, please elaborate

**Comment by renesansz** [comment]
@asafco4 , what we basically mean, is that could you verify that the `app.js` file you are using is the right one being used by your app?

Just try to put anything to indicate that the changes from your `js` file was being updated in the app.
I know this could be silly, but sometimes I do commit this kind of mistake, wherein I was very frustrated on how I keep on debugging the app and I've realized I was modifying the old copy of the app and I was debugging the new one, so I always keep wondering why I still receive the same error even though I keep on changing my code.

Because currently, there's no problem with how you've implemented your initSession..
Maybe it's just that your js file wasn't included in the app upon build.

**Comment by asafco4** [comment]
@renesansz  - I tried it: I added new debug messages for the onDeviceReady(), and I can see them, meaning file is indeed being used.

@tkuben  - wanna eleborate on it? I didn't understand.
I tried to touch an empty file to root directory of the app, it didn't effect anything.

**Comment by tkuben** [comment]
Hi @asafco4 

Are you familiar with debugging an android app using chrome? Please follow these steps:

https://developer.chrome.com/devtools/docs/remote-debugging

Once you have the chrome developer tools loaded. Go into the Network Tab. Click on JS and reload the app. Do you see any "red" files? or a file called branch.js that is being loaded from the root of your application? please take a screen shot of the js libs that are being loaded and paste in here. 

Cheers, 
TJ

**Comment by asafco4** [comment]
Yes, that's the tool I work with.
The error I see is in red:


app.js:95 Branch NEW onDeviceReady()
app.js:97 Branch NEW onDeviceReady ended...
ping:1 Given URL is not allowed by the Application configuration: One or
more of the given URLs is not allowed by the App's settings.  It must match
the Website URL or Canvas URL, or the domain must be a subdomain of one of
the App's domains.
flash_login.js:42 getAppName:  Object
flash_login.js:45 getSupportEmail:  Object
login.js:104 LoginController: User ID:  undefined
index.html#/login:1 Uncaught ReferenceError: DeepLinkHandler is not defined
app.js:89 Branch onResume()
app.js:91 Branch Resumed...
undefined:1 Uncaught (in promise) ReferenceError: DeepLinkHandler is not
defined(…)

**Comment by renesansz** [comment]
@asafco4 ,
can you post your source code of your app js?

Thanks

**Comment by asafco4** [comment]
@renesansz  - I sent it to your support email.
Anything you can do - please do!

**Comment by renesansz** [comment]
Hey @asafco4 ,
Just have received your email. In your source code, `DeepLinkHandler` should be defined as a global function not inside your Angular app.

Could you try to refactor your code and see if it works?
Let me know if this resolves your problem.

Cheers,
Rene

**Comment by asafco4** [comment]
Do you know where should I put it so it will be global? I tried to search
the Web,  also posted a question in stackoverflow.com but no luck yet.

**Comment by renesansz** [comment]
@asafco4 ,
What I mean by global is just to put the function outside the Angular scope.
Or just try to create a `script` tag.

```
<script>
function DeepLinkHandler(data) {
     console.log(data);
}
</script>
```

**Comment by asafco4** [comment]
I did this:
<script type="text/javascript" src="js/app/branch.js"></script>

and in branch.js I inserted the function and it's working!

You're the best, thanks a lot for all the help and patience!!!

Edit the instructions for Ionic/Cordova and add the requirement that this
function must be global + how to do it.

**Comment by renesansz** [comment]
cool! :+1: ,
No problem, I'm glad that it works now.
Actually we have it defined in our [README](https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK#initsession).
We'll revise a bit the instruction to make it clear for others, thank you for the suggestion.

Closing the issue now.

**Comment by barocsi** [comment]
Defining a global deeplink handler is ok. But should EXPLICITLY define in the docs as well.

**Comment by barocsi** [comment]
Still not working while compiling...
```
No resource identifier found for attribute 'autoVerify' in package 'android'
```

**Comment by barocsi** [comment]
I had to change 
/platforms/android/project.properties
from target=22  to target=23
however this is supposed to be a generated file, what generates it and based on that what, what should I change to make this change persistent.

**Comment by aeromusek** [comment]
@aaustin @renesansz could one of you clarify the correct process here for @barocsi?

(Reopening issue)

**Comment by renesansz** [comment]
@aeromusek ,
sure, will check on it 👌

**Comment by renesansz** [comment]
@barocsi ,
What version of the plugin are you using? Please also provide the Cordova version & list of plugins that you are using.
Currently, I can't replicate the issue, I've tried using the testbed app and it seems it defines the `target` value to `android-23`.

**Comment by barocsi** [comment]
I have cloned the newest Branch SDK from this repo as per doc, 
using Cordova 6.0.0, Cordova Android platform version 4.1.1.
I was using the previous version for Branch.io, removed the plugin and re-added it, and it did not change the target value, I had to do it manually, however since then, consequent rebuilds did not affect it. I am not sure what should trigger or lock the version update in the project.properties file.

Here is the list of my plugins,
```
com.raananw.imageResizer 0.1.2 "ImageResizer"
com.sarriaroman.PhotoViewer 1.1.1 "PhotoViewer"
com.verso.cordova.clipboard 0.1.0 "Clipboard"
cordova-plugin-camera 2.1.1 "Camera"
cordova-plugin-console 1.0.2 "Console"
cordova-plugin-crosswalk-webview 1.5.0 "Crosswalk WebView Engine"
cordova-plugin-device 1.1.1 "Device"
cordova-plugin-facebook4 1.7.0 "Facebook Connect"
cordova-plugin-file 3.0.0 "File"
cordova-plugin-google-analytics 0.8.1 "Google Universal Analytics Plugin"
cordova-plugin-inappbrowser 1.2.1 "InAppBrowser"
cordova-plugin-network-information 1.2.0 "Network Information"
cordova-plugin-sim 1.2.0 "SIM"
cordova-plugin-splashscreen 3.2.0 "Splashscreen"
cordova-plugin-whitelist 1.2.1 "Whitelist"
cordova-plugin-x-socialsharing 5.0.9 "SocialSharing"
es6-promise-plugin 3.0.2 "Promise"
io.branch.sdk 2.0.11 "branch-cordova-sdk"
ionic-plugin-keyboard 2.0.1 "Keyboard"
phonegap-plugin-push 1.5.3 "PushPlugin"
```

**Comment by aaustin** [comment]
If your build target is not 23, you'll get this error. This silently removes the autoVerify tag if the build target is < 23. https://github.com/BranchMetrics/Cordova-Ionic-PhoneGap-Deferred-Deep-Linking-SDK/pull/135

**Comment by ouihamdaoui** [comment]
Log For : app.apk 

 Log Type : Recompiling 

 Log Recorded At : 12 2018-01- 2:22:58,11 

 Log Recorded By : Advanced ApkTool v4.2.0 By BDFreak 

------------------------------------------------------ 

W: C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\AndroidManifest.xml:16: error: No resource identifier found for attribute 'roundIcon' in package 'android'
W: 
Exception in thread "main" brut.androlib.AndrolibException: brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = 1): [C:\Users\pc\AppData\Local\Temp\brut_util_Jar_1479779000625798190.tmp, p, --forced-package-id, 127, --min-sdk-version, 15, --target-sdk-version, 26, --version-code, 10, --version-name, 10.0, --no-version-vectors, -F, C:\Users\pc\AppData\Local\Temp\APKTOOL5360200212280990529.tmp, -0, arsc, -0, arsc, -I, C:\Users\pc\Desktop\AdvancedApkTool\1-BDFreak\Frameworks\1.apk, -S, C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\res, -M, C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\AndroidManifest.xml]
	at brut.androlib.Androlib.buildResourcesFull(Androlib.java:492)
	at brut.androlib.Androlib.buildResources(Androlib.java:426)
	at brut.androlib.Androlib.build(Androlib.java:305)
	at brut.androlib.Androlib.build(Androlib.java:270)
	at brut.apktool.Main.cmdBuild(Main.java:227)
	at brut.apktool.Main.main(Main.java:75)
Caused by: brut.androlib.AndrolibException: brut.common.BrutException: could not exec (exit code = 1): [C:\Users\pc\AppData\Local\Temp\brut_util_Jar_1479779000625798190.tmp, p, --forced-package-id, 127, --min-sdk-version, 15, --target-sdk-version, 26, --version-code, 10, --version-name, 10.0, --no-version-vectors, -F, C:\Users\pc\AppData\Local\Temp\APKTOOL5360200212280990529.tmp, -0, arsc, -0, arsc, -I, C:\Users\pc\Desktop\AdvancedApkTool\1-BDFreak\Frameworks\1.apk, -S, C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\res, -M, C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\AndroidManifest.xml]
	at brut.androlib.res.AndrolibResources.aaptPackage(AndrolibResources.java:456)
	at brut.androlib.Androlib.buildResourcesFull(Androlib.java:478)
	... 5 more
Caused by: brut.common.BrutException: could not exec (exit code = 1): [C:\Users\pc\AppData\Local\Temp\brut_util_Jar_1479779000625798190.tmp, p, --forced-package-id, 127, --min-sdk-version, 15, --target-sdk-version, 26, --version-code, 10, --version-name, 10.0, --no-version-vectors, -F, C:\Users\pc\AppData\Local\Temp\APKTOOL5360200212280990529.tmp, -0, arsc, -0, arsc, -I, C:\Users\pc\Desktop\AdvancedApkTool\1-BDFreak\Frameworks\1.apk, -S, C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\res, -M, C:\Users\pc\Desktop\AdvancedApkTool\3-Out\app.apk\AndroidManifest.xml]
	at brut.util.OS.exec(OS.java:95)
	at brut.androlib.res.AndrolibResources.aaptPackage(AndrolibResources.java:450)
	... 6 more
